[Tan Ying Hui] iP

ureshiiYing opened the pull request.

Commit:
Add Gradle support

ureshiiYing committed:
Working on Duke Level 1

ureshiiYing committed:
Trying out the text ui testing

ureshiiYing committed:
Found a bug and retested level 1

ureshiiYing committed:
Implemented Level 2 and did automated testing

ureshiiYing committed:
Implemented and tested level 3. Figured out the white spaces bug that kept ruining the tests

ureshiiYing committed:
Implemented level 4

ureshiiYing committed:
Tested everything done so far with correct cases

ureshiiYing committed:
Implementing level 5

Used exceptions and if, else and print statements to signal and display error
Reformatted some of the if-else code to switch cases
Figured out a better way to split the input

ureshiiYing committed:
Implemented and tested level 5

Handled errors such as insufficient arguments in command, unknown commands, empty description for tasks
Did regression testing and passed
Added new test cases for error handling

ureshiiYing committed:
Minor edit to storing the time for these task subclasses

Removed any white spaces in front/ behind the input
Also passed the same test cases as before

ureshiiYing committed:
Implemented and tested Level 6

Added the delete feature and handled the same type of errors handled previously
Did regression testing and passed
Added new delete test cases

ureshiiYing committed:
Minor edit to change the response for delete command errors

ureshiiYing committed:
Implemented Level 7

Added a SaveLoad class that is responsible for loading and saving a local save file for PandaBot
Handled error for any loading error in the SaveLoad class

ureshiiYing committed:
Implemented Level 8

Created a DateAndTime class which allows PandaBot to understand date and time in the format dd/mm/yyyy hhmm
Edited the Deadline behaviour such that Deadlines can accept both formatted and unformatted input as due by dates

ureshiiYing committed:
Merge branch 'branch-Level-7'

ureshiiYing committed:
Merge branch 'branch-Level-8'

ureshiiYing committed:
Refactor the code to be more OOP

ureshiiYing committed:
Added JUnit Tests

ureshiiYing committed:
Implemented Level 9

Added FindCommand feature

ureshiiYing committed:
Writing JavaDoc

ureshiiYing committed:
Write JavaDoc for all the classes

ureshiiYing committed:
Tweak code to comply with CS2103T coding standard

ureshiiYing committed:
Merge branch 'branch-Level-9'

ureshiiYing committed:
Merge branch 'branch-A-JavaDoc'

ureshiiYing committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/DeleteCommand.java
#	src/main/java/Event.java
#	src/main/java/PandaBot.java
#	src/main/java/PandaBotException.java
#	src/main/java/PandaBotInvalidArgumentFormatException.java
#	src/main/java/PandaBotInvalidCommandException.java
#	src/main/java/PandaBotLoadingTasksErrorException.java

ureshiiYing committed:
Edit JavaDoc comments, reformat some code to fit coding standard

ureshiiYing committed:
Tweak the logic of the code regarding date input

Missed out checking the validity of the date input, which has been added

UnicornJin commented on src/main/java/AddCommand.java:
```diff
@@ -0,0 +1,36 @@
+        this.task = task;
+    }
+
+    /**
```
Good JavaDoc

UnicornJin commented on src/main/java/ByeCommand.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Represents a bye command which allows users to exit the program.
+ */
+public class ByeCommand extends Command {
```
It seems you are using different classes for UI output, it's okay.
You can also do this in one single class lah

ureshiiYing replied:
I thought it would be better if the Ui class was in charge of all the messages to be printed (Single Responsibility Principle?)

UnicornJin commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,27 @@
+ * Represents a Command, which is an abstract class.
+ * The Command class is used to execute a command.
+ */
+public abstract class Command {
```
Oh you use abstract class, that's good.

UnicornJin commented on src/main/java/FindCommand.java:
```diff
@@ -0,0 +1,34 @@
+     * @throws PandaBotException If any errors occurs when executing the command
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws PandaBotException {
```
Maybe name this function as executeOrder/executeCommand would be better

UnicornJin reviewed (commented):
Nice code, Jia You

JinHao-L commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+    private String dueBy;
+
+    /**
+     * Create a Deadline object.
```
Should it be "Creates a Deadline object" instead?

ureshiiYing replied:
Yes, thanks for pointing that out.

JinHao-L commented on src/main/java/ByeCommand.java:
```diff
@@ -0,0 +1,29 @@
+     * Otherwise, returns false if the program should continue to run.
+     * 
+     * @return true as the program exits on bye command
+     */
```
I think that the {@inheritDoc} might be helpful here if the JavaDoc of the overridden method is the same.

JinHao-L commented on src/main/java/PandaBotEmptyTaskDescriptionException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Represents a type of PandaBotException which is thrown when the task description is empty.
+ */
+public class PandaBotEmptyTaskDescriptionException extends PandaBotException{
```
Should there be whitespace before the curly braces?

ureshiiYing replied:
Yes, thanks for pointing that out.

JinHao-L commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+    }
+    
+    private Task convertToTask(String input) throws PandaBotException{
+        String[] tDes = input.split(" \\| ");
```
Perhaps a more intuitive variable name here? For example, taskDetails

JinHao-L commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,139 @@
+                    break;
+                }
+            }
+        }
```
Just a thought of mine but perhaps these implementations could be handled by their respective command and Ui could purely be in charge of printing out the String outputs given by the execution of the commands

ureshiiYing replied:
I realised your suggestion came in handy, for the GUI implementation. Thanks!

JinHao-L reviewed (commented):
Your code is easy to follow and understand. I like how you separated the commands into the different command type as this allows for easy addition of more command type in the future and also customisation of each command execution. The structure of your program was good and really complies with OOP. 

One suggestion that I would like to make is to use packaging as it helps structure and organises your project files, especially when your project files grow with each iteration.

Overall, the project is quite well done and a pleasure to read. All the best!! 👍

ureshiiYing committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ureshiiYing committed:
Set up Gradle

Set up Gradle such that it can build and run PandaBot, and run the unit tests
Minor Edit to code to follow coding standard as pointed out in peer review

ureshiiYing committed:
Reformat code after using checkstyle

nicktohzyu commented on src/main/java/DateAndTime.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class DateAndTime {
+    private final LocalDate date;
+    private final LocalTime time;
```
Consider using LocalDateTime instead.

nicktohzyu commented on src/main/java/DateAndTime.java:
```diff
@@ -0,0 +1,65 @@
+
+    /**
+     * Creates a DateAndTime object by taking in an input
+     * of the format dd/mm/yyyy hhmm.
```
Look up ISO 8601.

nicktohzyu commented on src/main/java/DateAndTime.java:
```diff
@@ -0,0 +1,65 @@
+        this.time = LocalTime.parse(time, DateTimeFormatter.ofPattern("HHmm"));
+    }
+
+    private String parseDate(LocalDate input) {
```
There probably is an existing API method for this

nicktohzyu commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+        String[] dT = input.split(" ");
+        if (dT.length > 1) {
+            try {
+                DateAndTime dateTime = new DateAndTime(dT[0], dT[1]);
```
This logic should be in DateAndTime (if you choose to keep that class), especially because this is the only place the constructor is used

nicktohzyu commented on src/main/java/PandaBot.java:
```diff
@@ -0,0 +1,52 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCmd();
+                ui.printLine(); // show the divider line ("_______")
```
The word "line" is ambiguous, as in system.out.println. Maybe call it printDivider instead?

nicktohzyu commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    private static String[] obtainDes(String des, String separator) throws PandaBotInsufficientArgumentException {
```
Meaning of "des" should be explained.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        // create a dir for save files to be saved in
+        if (!directoryExists) {
+            try {
+                java.nio.file.Files.createDirectory(dirPath);
```
`createDirectory` already checks if the directory exists.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        // create a save file where the data will be written to
+        if (!saveFileExists) {
+            try {
+                Files.createFile(saveFilePath);
```
Same comment as `createDirectory`.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        return tasks;
+    }
+
+    private void loadTask(ArrayList<Task> tasks, BufferedReader reader) throws IOException {
```
Perhaps this should be called `loadTasks`, as it can load multiple.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @return a list of tasks as an ArrayList
+     */
+    public ArrayList<Task> load() {
```
Seems like loadTask can be absorbed into this function.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+}
```
Consider using serialization instead. Otherwise, good practice would be to write tests for the conversion methods here.

nicktohzyu commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * @throws PandaBotEmptyTaskDescriptionException If the description given is empty
+     */
+    public Task(String description) throws PandaBotEmptyTaskDescriptionException {
+        this.description = description.strip(); // removes starting and ending white spaces
```
Redundant comment, there's no need to explain what strip does (it has its own javadocs).

nicktohzyu commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * @return a String that represents a status icon which marks whether the task is done
+     */
+    public String getStatusIcon() {
+        return (isDone ? "\u2713" : "\u2718"); //return tick or X symbols
```
Nice! Could be absorbed into `toString` as both methods are short.

nicktohzyu commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+
+    /**
+     * Creates a TaskList object that stores the tasks.
+     * @param taskList an ArrayList of tasks to be stored
```
Should add a blank line between description and param.

nicktohzyu commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    public Task getTaskAt(int index) throws PandaBotOutOfRangeException {
+        try {
+            return taskList.get(index);
+        } catch (IndexOutOfBoundsException e) {
```
Great use of try-catch!

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        Path saveFilePath = Paths.get(home, "PandaBot", fileName);
```
Shadows line 14.

ureshiiYing replied:
Ah yes it did, thanks for pointing that out

ureshiiYing replied:
Ah yes it did, thanks for catching that.

Update: I realised I created that Path and set the path field to that in a later line, so it did not shadow. Thanks anyways.

nicktohzyu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,130 @@
+ * loading and saving tasks into the save file.
+ */
+public class Storage {
+    private Path saveFilePath;
```
Can be final.

ureshiiYing replied:
Thanks for all the suggestions and catching the errors!

nicktohzyu reviewed (commented):
General advice:
Group classes into packages. A worse but easier alternative would be to organize them in folders. There are too many classes in the base directory.

Consider serialization API for data storage instead of manual parsing (more prone to error).

Wholesome goodbye message! haha

ureshiiYing committed:
Organise the classes into packages

ureshiiYing committed:
Add a GUI for PandaBot

ureshiiYing committed:
Add  javadocs for new methods and classes and ran checkstyle

Previous commit did not include javadocs and some of the code do not adhere to the coding standard.
For documentation purposes, include javadocs for new methods and classes and update them for the other methods and classes.
To ensure that the coding standard was adhered to, run and pass checkstyle.

ureshiiYing committed:
Merge branch 'branch-Level-10'

ureshiiYing committed:
Update build.gradle file

ureshiiYing committed:
Implement A-Assertions

Currently, no assertions are used.
Assertions can be used to verify any assumptions made while coding.
This differs from using exceptions to check external inputs.

Use assertions to the code verify that any assumptions made holds true.

ureshiiYing committed:
Improve Code Quality

Some parts of the code do not follow the recommended code quality guidelines.

Code should be of high quality in order to maximise readibility.

Improve code quality by following the recommended guidelines.

ureshiiYing committed:
Refactor code to use LocalDateTime class instead

ureshiiYing committed:
Improve code quality by using proper naming

Some of the method and variable names are not well-named.
This could lead to potential confusion.
Use proper names and use names to explain the code to improve code quality.

ureshiiYing committed:
Merge pull request #1 from ureshiiYing/branch-A-Assertions

Implement A-Assertions

ureshiiYing committed:
Merge branch 'master' of https://github.com/ureshiiYing/ip

ureshiiYing committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/pandabot/parser/Parser.java
#	src/main/java/pandabot/storage/Storage.java

ureshiiYing committed:
Merge pull request #2 from ureshiiYing/branch-A-CodeQuality

Improve Code Quality

ureshiiYing committed:
Merge branch 'master' of https://github.com/ureshiiYing/ip

ureshiiYing committed:
Set up Continuous Integration (CI)

ureshiiYing committed:
Merge branch 'master' of https://github.com/ureshiiYing/ip

ureshiiYing committed:
Update Parser.java to follow CS2103 coding standard

ureshiiYing committed:
Update gradle.yml to exclude I/O redirection tests

ureshiiYing committed:
Add support for managing tasks to be done after a specific time or task.

This gives PandaBot more features.

An additional change is: Extracting a parent class, TimedTask.

Tasks such as Event, Deadline, and the newly added DoAfter have a common attribute - time.
Processing the various tasks with time separately causes code duplication.

Let's extract the time attribute into a new parent class named TimedTask, which can be used to
represent tasks with time attribute. It can accept both unformatted and formatted time input.

ureshiiYing committed:
Merge branch 'branch-B-DoAfterTasks'

ureshiiYing committed:
Add help command

PandaBot does not have any means to allow users to be sure that they are giving the correct input format.

This could confuse users who are not familiar with PandaBot's features.

Add a help command which will bring up a help page that contains a list of commands that the
user can use.

Let's make this help page to help familiarise users with PandaBot's available commands.

ureshiiYing committed:
Improve GUI to make it more polished

The PandaBot GUI is very basic.

It can be improved by personalising the GUI.

Let's
* Make the GUI window responsive to resizing of window
* Add a background to PandaBot using CSS
* Stylize the dialog box using CSS

ureshiiYing committed:
Rename user image

ureshiiYing committed:
Merge branch 'A-BetterGui'

ureshiiYing committed:
Add a User Guide

ureshiiYing committed:
Set theme jekyll-theme-minimal

ureshiiYing committed:
Set theme jekyll-theme-slate

ureshiiYing committed:
Set theme jekyll-theme-cayman

ureshiiYing committed:
Set theme jekyll-theme-minimal

ureshiiYing committed:
Set theme jekyll-theme-midnight

ureshiiYing committed:
Set theme jekyll-theme-minimal

ureshiiYing committed:
Set theme jekyll-theme-cayman

ureshiiYing committed:
Merge branch 'master' of https://github.com/ureshiiYing/ip

ureshiiYing committed:
Update User Guide
